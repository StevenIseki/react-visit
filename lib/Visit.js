'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Visit = function (_Component) {
  _inherits(Visit, _Component);

  function Visit(props) {
    _classCallCheck(this, Visit);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Visit).call(this, props));

    _this.state = {
      visited: false
    };
    return _this;
  }

  _createClass(Visit, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener('scroll', this.handleScroll.bind(this));
    }
  }, {
    key: 'handleScroll',
    value: function handleScroll(e) {
      if (this.refs.visit != undefined) {

        if (this.isElementInViewport(this.refs.visit)) {
          if (this.state.visited) {
            /* already visited */
          } else {
              /* fire off event for element visited */
              this.props.visited();
              this.setState({ visited: true });
            }
        }
      }
    }
  }, {
    key: 'isElementInViewport',
    value: function isElementInViewport(el) {
      var rect = el.getBoundingClientRect();
      return rect.top >= 0 && rect.left >= 0 && rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) && rect.right <= (window.innerWidth || document.documentElement.clientWidth);
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement('span', { ref: 'visit' });
    }
  }]);

  return Visit;
}(_react.Component);

exports.default = Visit;