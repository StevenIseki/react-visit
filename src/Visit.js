import React, { Component } from 'react';

export default class Visit extends Component {

  constructor(props) {
    super(props);

    this.state = {
      visited: false
    }
  }

  componentDidMount() {
    window.addEventListener('scroll', this.handleScroll.bind(this))
  }

  handleScroll (e) {
    if(this.refs.visit != undefined){

      if(this.isElementInViewport(this.refs.visit)){
        if(this.state.visited){
          /* already visited */
        } else {
          /* fire off event for element visited */
          this.props.visited()
          this.setState({ visited: true });
        }
      }
    }
  }

  isElementInViewport (el) {
      var rect = el.getBoundingClientRect();
      return (
          rect.top >= 0 &&
          rect.left >= 0 &&
          rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
          rect.right <= (window.innerWidth || document.documentElement.clientWidth)
      );
  }

  render() {
    return (
    	<span ref='visit' />
    )
  }
}
