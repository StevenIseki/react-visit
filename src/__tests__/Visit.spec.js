/* setup enzyme */
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
configure({ adapter: new Adapter() })

/* setup jsdom */
var jsdom = require('jsdom')
const { JSDOM } = jsdom
const window = new JSDOM('').window
global.window = window
global.document = window.document

import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Visit from '../Visit'

test('Visit renders correctly and matches snapshot', () => {
  const handleVisit = jest.fn()
  const handleLeave = jest.fn()

  let visitStyle = {
    position: 'absolute',
    visibility: 'hidden',
    width: '100%',
    marginTop: '-10rem',
    height: '10rem'
 }

  const component = renderer.create(
    <Visit
      visitStyle={visitStyle}
      onVisit={handleVisit}
      onLeave={handleLeave} />
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Visit renders the correct elements and props', () => {
  const handleVisit = jest.fn()
  const handleLeave = jest.fn()

  let visitStyle = {
    position: 'absolute',
    visibility: 'hidden',
    width: '100%',
    marginTop: '-10rem',
    height: '10rem'
 }

 const wrapper = shallow(
    <Visit
      className='visit'
      visitStyle={visitStyle}
      onVisit={handleVisit}
      onLeave={handleLeave} />
  )

  expect(wrapper.instance().props.visitStyle).toEqual(visitStyle)
  expect(wrapper.instance().props.onVisit).toEqual(handleVisit)
  expect(wrapper.instance().props.onLeave).toEqual(handleLeave)
  expect(wrapper.instance().props.className).toEqual('visit')

  expect(wrapper.find('span').length).toEqual(1)
  expect(wrapper.find('span').text()).toEqual('')
  expect(wrapper.find('span').props().className).toEqual('visit')

  setTimeout(() => {
    expect(handleVisit).toBeCalled();
  }, 0)

  // console.log(wrapper.debug())
})
