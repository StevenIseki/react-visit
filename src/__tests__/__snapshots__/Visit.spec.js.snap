// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visit renders correctly and matches snapshot 1`] = `
<span
  style={
    Object {
      "height": "10rem",
      "marginTop": "-10rem",
      "position": "absolute",
      "visibility": "hidden",
      "width": "100%",
    }
  }
/>
`;
